google/resumable_media/_helpers.py

```diff
@@ -162,23 +162,43 @@ def wait_and_retry(func, get_status_code, retry_strategy):
     Returns:
         object: The return value of ``func``.
     """
-    response = func()
-    if get_status_code(response) not in RETRYABLE:
-        return response
-
     total_sleep = 0.0
     num_retries = 0
     base_wait = 0.5  # When doubled will give 1.0
-    while retry_strategy.retry_allowed(total_sleep, num_retries):
+
+    # Set the retriable_exception_type if possible. We expect requests to be
+    # present here and the transport to be using requests.exceptions errors,
+    # but due to loose coupling with the transport layer we can't guarantee it.
+    try:
+        import requests
```

Comment:
This module is "above" the transport-specific `requests` subpackage, which breaks the "alternate transports are possible" design goal.

Because `google.resumable_media._helpers.wait_and_retry` is only called from one place, `google.resumable_media.requests._request_helpers.http_request`, it seems like we could just move `wait_and_retry` into that module, which would remove the need for the conditional import, as well as preserving the transport-independence of `google.resumable_media._helpers`.

If so, we should also move `google._async_resumable_media._helpers.wait_and_retry` down into `google._async_resumable_media.requests._request_helpers` as well, to maintain the (really painfully duplicated) structure.

Reply:
The try/except import model here will guarantee alternative transports work, though since ConnectionError is specific to the transport, connection errors can't be caught on alternative transports. Given that the only thing changing is awareness of connection errors, I'm not sure I agree there's value to moving this into http_request, unless http_request is where it should have been all along. All processing of HTTP response codes is in this file and not in http_request so it makes sense to keep it here.

Reply:
The requests import is nominally temporary as Python 3 standardizes on ConnectionError (stdlib), but no idea when that will really be the standard in practice as requests does not support it yet.